// @metamask/snaps-cli/dist/index.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeWithRules = exports.mergeWithCustomize = exports.merge = exports.SnapsBundleWarningsPlugin = exports.SnapsBuiltInResolver = exports.SnapsWatchPlugin = exports.SnapsStatsPlugin = exports.getDefaultConfiguration = exports.utils = void 0;
exports.utils = __importStar(require("./utils/index.cjs"));
var webpack_1 = require("./webpack/index.cjs");
Object.defineProperty(exports, "getDefaultConfiguration", { enumerable: true, get: function () { return webpack_1.getDefaultConfiguration; } });
Object.defineProperty(exports, "SnapsStatsPlugin", { enumerable: true, get: function () { return webpack_1.SnapsStatsPlugin; } });
Object.defineProperty(exports, "SnapsWatchPlugin", { enumerable: true, get: function () { return webpack_1.SnapsWatchPlugin; } });
Object.defineProperty(exports, "SnapsBuiltInResolver", { enumerable: true, get: function () { return webpack_1.SnapsBuiltInResolver; } });
Object.defineProperty(exports, "SnapsBundleWarningsPlugin", { enumerable: true, get: function () { return webpack_1.SnapsBundleWarningsPlugin; } });
// Re-exported from `snaps-cli` for convenience.
var webpack_merge_1 = require("webpack-merge");
Object.defineProperty(exports, "merge", { enumerable: true, get: function () { return webpack_merge_1.merge; } });
Object.defineProperty(exports, "mergeWithCustomize", { enumerable: true, get: function () { return webpack_merge_1.mergeWithCustomize; } });
Object.defineProperty(exports, "mergeWithRules", { enumerable: true, get: function () { return webpack_merge_1.mergeWithRules; } });
