// @metamask/permission-controller/dist/rpc-methods/requestPermissions.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.requestPermissionsHandler = void 0;
const controller_utils_1 = require("@metamask/controller-utils");
const errors_1 = require("../errors.cjs");
const utils_1 = require("../utils.cjs");
exports.requestPermissionsHandler = {
    methodNames: [utils_1.MethodNames.RequestPermissions],
    implementation: requestPermissionsImplementation,
    hookNames: {
        requestPermissionsForOrigin: true,
    },
};
/**
 * Request Permissions implementation to be used in JsonRpcEngine middleware.
 *
 * @param req - The JsonRpcEngine request
 * @param res - The JsonRpcEngine result object
 * @param _next - JsonRpcEngine next() callback - unused
 * @param end - JsonRpcEngine end() callback
 * @param options - Method hooks passed to the method implementation
 * @param options.requestPermissionsForOrigin - The specific method hook needed for this method implementation
 * @returns A promise that resolves to nothing
 */
async function requestPermissionsImplementation(req, res, _next, end, { requestPermissionsForOrigin }) {
    const { params } = req;
    if (!Array.isArray(params) || !(0, controller_utils_1.isPlainObject)(params[0])) {
        return end((0, errors_1.invalidParams)({ data: { request: req } }));
    }
    const [requestedPermissions] = params;
    const [grantedPermissions] = await requestPermissionsForOrigin(requestedPermissions);
    // `wallet_requestPermission` is specified to return an array.
    res.result = Object.values(grantedPermissions);
    return end();
}
