// @metamask/permission-controller/dist/rpc-methods/revokePermissions.mjs
import { isNonEmptyArray } from "@metamask/utils";
import { invalidParams } from "../errors.mjs";
import { MethodNames } from "../utils.mjs";
export const revokePermissionsHandler = {
    methodNames: [MethodNames.RevokePermissions],
    implementation: revokePermissionsImplementation,
    hookNames: {
        revokePermissionsForOrigin: true,
    },
};
/**
 * Revoke Permissions implementation to be used in JsonRpcEngine middleware.
 *
 * @param req - The JsonRpcEngine request
 * @param res - The JsonRpcEngine result object
 * @param _next - JsonRpcEngine next() callback - unused
 * @param end - JsonRpcEngine end() callback
 * @param options - Method hooks passed to the method implementation
 * @param options.revokePermissionsForOrigin - A hook that revokes given permission keys for an origin
 * @returns A promise that resolves to nothing
 */
async function revokePermissionsImplementation(req, res, _next, end, { revokePermissionsForOrigin }) {
    const { params } = req;
    const param = params?.[0];
    if (!param) {
        return end(invalidParams({ data: { request: req } }));
    }
    // For now, this API revokes the entire permission key
    // even if caveats are specified.
    const permissionKeys = Object.keys(param);
    if (!isNonEmptyArray(permissionKeys)) {
        return end(invalidParams({ data: { request: req } }));
    }
    revokePermissionsForOrigin(permissionKeys);
    res.result = null;
    return end();
}
