// @metamask/snaps-utils/dist/handlers.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnNameLookupResponseStruct = exports.DomainResolutionResponseStruct = exports.AddressResolutionResponseStruct = exports.DomainResolutionStruct = exports.AddressResolutionStruct = exports.OnHomePageResponseStruct = exports.OnHomePageResponseWithIdStruct = exports.OnHomePageResponseWithContentStruct = exports.OnSignatureResponseStruct = exports.OnTransactionResponseStruct = exports.OnTransactionResponseWithContentStruct = exports.OnTransactionResponseWithIdStruct = exports.OnTransactionSeverityResponseStruct = exports.SNAP_EXPORTS = void 0;
const snaps_sdk_1 = require("@metamask/snaps-sdk");
const superstruct_1 = require("@metamask/superstruct");
const handler_types_1 = require("./handler-types.cjs");
exports.SNAP_EXPORTS = {
    [handler_types_1.HandlerType.OnRpcRequest]: {
        type: handler_types_1.HandlerType.OnRpcRequest,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnTransaction]: {
        type: handler_types_1.HandlerType.OnTransaction,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnCronjob]: {
        type: handler_types_1.HandlerType.OnCronjob,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnNameLookup]: {
        type: handler_types_1.HandlerType.OnNameLookup,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnInstall]: {
        type: handler_types_1.HandlerType.OnInstall,
        required: false,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnUpdate]: {
        type: handler_types_1.HandlerType.OnUpdate,
        required: false,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnKeyringRequest]: {
        type: handler_types_1.HandlerType.OnKeyringRequest,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnHomePage]: {
        type: handler_types_1.HandlerType.OnHomePage,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnSignature]: {
        type: handler_types_1.HandlerType.OnSignature,
        required: true,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
    [handler_types_1.HandlerType.OnUserInput]: {
        type: handler_types_1.HandlerType.OnUserInput,
        required: false,
        validator: (snapExport) => {
            return typeof snapExport === 'function';
        },
    },
};
exports.OnTransactionSeverityResponseStruct = (0, superstruct_1.object)({
    severity: (0, superstruct_1.optional)((0, superstruct_1.literal)(snaps_sdk_1.SeverityLevel.Critical)),
});
exports.OnTransactionResponseWithIdStruct = (0, superstruct_1.assign)(exports.OnTransactionSeverityResponseStruct, (0, superstruct_1.object)({
    id: (0, superstruct_1.string)(),
}));
exports.OnTransactionResponseWithContentStruct = (0, superstruct_1.assign)(exports.OnTransactionSeverityResponseStruct, (0, superstruct_1.object)({
    content: snaps_sdk_1.ComponentOrElementStruct,
}));
exports.OnTransactionResponseStruct = (0, superstruct_1.nullable)((0, superstruct_1.union)([
    exports.OnTransactionResponseWithContentStruct,
    exports.OnTransactionResponseWithIdStruct,
]));
exports.OnSignatureResponseStruct = exports.OnTransactionResponseStruct;
exports.OnHomePageResponseWithContentStruct = (0, superstruct_1.object)({
    content: snaps_sdk_1.ComponentOrElementStruct,
});
exports.OnHomePageResponseWithIdStruct = (0, superstruct_1.object)({
    id: (0, superstruct_1.string)(),
});
exports.OnHomePageResponseStruct = (0, superstruct_1.union)([
    exports.OnHomePageResponseWithContentStruct,
    exports.OnHomePageResponseWithIdStruct,
]);
exports.AddressResolutionStruct = (0, superstruct_1.object)({
    protocol: (0, superstruct_1.string)(),
    resolvedDomain: (0, superstruct_1.string)(),
});
exports.DomainResolutionStruct = (0, superstruct_1.object)({
    protocol: (0, superstruct_1.string)(),
    resolvedAddress: (0, superstruct_1.string)(),
    domainName: (0, superstruct_1.string)(),
});
exports.AddressResolutionResponseStruct = (0, superstruct_1.object)({
    resolvedDomains: (0, superstruct_1.size)((0, superstruct_1.array)(exports.AddressResolutionStruct), 1, Infinity),
});
exports.DomainResolutionResponseStruct = (0, superstruct_1.object)({
    resolvedAddresses: (0, superstruct_1.size)((0, superstruct_1.array)(exports.DomainResolutionStruct), 1, Infinity),
});
exports.OnNameLookupResponseStruct = (0, superstruct_1.nullable)((0, superstruct_1.union)([exports.AddressResolutionResponseStruct, exports.DomainResolutionResponseStruct]));
